File: io/json/JsonEncoder.java
package common.io.json;

import java.lang.reflect.Array;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonNull;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;

import common.io.json.JsonClass.JCGeneric;
import common.io.json.JsonClass.JCGenericWrite;
import common.io.json.JsonException.Type;
import common.util.Data;

public class JsonEncoder {

	public static JsonElement encode(Object obj) {
		return Data.err(() -> encode(obj, null));
	}

	private static JsonElement encode(Object obj, JsonEncoder par) throws Exception {
		if (obj == null)
			return JsonNull.INSTANCE;
		if (obj instanceof JsonElement)
			return (JsonElement) obj;
		if (obj instanceof Number)
			return new JsonPrimitive((Number) obj);
		if (obj instanceof Boolean)
			return new JsonPrimitive((Boolean) obj);
		if (obj instanceof String)
			return new JsonPrimitive((String) obj);
		if (obj instanceof Class)
			return new JsonPrimitive(((Class<?>) obj).getName());
		Class<?> cls = obj.getClass();
		if (cls.isArray()) {
			if (par.curjfld != null && par.curjfld.usePool()) {
				JsonField.Handler handler = new JsonField.Handler();
				int n = Array.getLength(obj);
				JsonArray jarr = new JsonArray();
				for (int i = 0; i < n; i++)
					jarr.add(handler.add(Array.get(obj, i)));
				JsonObject jobj = new JsonObject();
				jobj.add("pool", encode(handler.list));
				jobj.add("data", jarr);
				return jobj;
			}
			int n = Array.getLength(obj);
			JsonArray arr = new JsonArray(n);
			for (int i = 0; i < n; i++)
				arr.add(encode(Array.get(obj, i), par));
			return arr;
		}
		if (cls.getAnnotation(JCGeneric.class) != null && par != null && par.curjfld.alias().length == 1) {
			JCGeneric jcg = cls.getAnnotation(JCGeneric.class);
			Class<?> alias = par.curjfld.alias()[0];
			boolean found = false;
			for (Class<?> ala : jcg.value())
				if (ala == alias) {
					found = true;
					break;
				}
			if (!found)
				throw new JsonException(Type.TYPE_MISMATCH, null, "class not present in JCGeneric");
			for (Method m : cls.getMethods()) {
				JCGenericWrite jcgw = m.getAnnotation(JCGenericWrite.class);
				if (jcgw != null && jcgw.value() == alias) {
					return encode(m.invoke(obj), par);
				}
			}
			throw new JsonException(Type.TYPE_MISMATCH, null, "no JCGenericWrite present");
		}
		if (par != null && par.curjfld != null) {
			JsonField jfield = par.curjfld;
			if (jfield.ser() == JsonField.SerType.FUNC) {
				if (jfield.serializer().length() == 0)
					throw new JsonException(Type.FUNC, null, "no serializer function");
				Method m = par.obj.getClass().getMethod(jfield.serializer(), cls);
				return encode(m.invoke(par.obj, obj));
			} else if (jfield.ser() == JsonField.SerType.CLASS) {
				JsonClass cjc = cls.getAnnotation(JsonClass.class);
				if (cjc == null || cjc.serializer().length() == 0)
					throw new JsonException(Type.FUNC, null, "no serializer function");
				String func = cjc.serializer();
				Method m = cls.getMethod(func);
				return encode(m.invoke(obj), null);
			}
		}
		JsonClass jc = cls.getAnnotation(JsonClass.class);
		if (jc != null)
			if (jc.write() == JsonClass.WType.DEF)
				return new JsonEncoder(par, obj).ans;
			else if (jc.write() == JsonClass.WType.CLASS) {
				if (jc.serializer().length() == 0)
					throw new JsonException(Type.FUNC, null, "no serializer function");
				String func = jc.serializer();
				Method m = cls.getMethod(func);
				return encode(m.invoke(obj), null);
			}
		if (obj instanceof List)
			return encodeList((List<?>) obj, par);
		if (obj instanceof Set)
			return encodeSet((Set<?>) obj, par);
		if (obj instanceof Map)
			return encodeMap((Map<?, ?>) obj, par);
		throw new JsonException(Type.UNDEFINED, null, "object " + obj + " not defined");
	}

	private static JsonElement encodeList(List<?> list, JsonEncoder par) throws Exception {
		if (par != null && par.curjfld != null && par.curjfld.usePool()) {
			JsonField.Handler handler = new JsonField.Handler();
			int n = list.size();
			JsonArray jarr = new JsonArray();
			for (int i = 0; i < n; i++)
				jarr.add(handler.add(list.get(i)));
			JsonObject jobj = new JsonObject();
			jobj.add("pool", encode(handler.list));
			jobj.add("data", jarr);
			return jobj;
		}
		JsonArray ans = new JsonArray(list.size());
		for (Object obj : list)
			ans.add(encode(obj, par));
		return ans;
	}

	private static JsonArray encodeMap(Map<?, ?> map, JsonEncoder par) throws Exception {
		JsonArray ans = new JsonArray(map.size());
		for (Entry<?, ?> obj : map.entrySet()) {
			JsonObject ent = new JsonObject();
			ent.add("key", encode(obj.getKey(), par));
			ent.add("val", encode(obj.getValue(), par));
			ans.add(ent);
		}
		return ans;
	}

	private static JsonArray encodeSet(Set<?> set, JsonEncoder par) throws Exception {
		JsonArray ans = new JsonArray(set.size());
		for (Object obj : set)
			ans.add(encode(obj, par));
		return ans;
	}

	private final JsonEncoder par;
	private final Object obj;
	private final JsonObject ans = new JsonObject();

	private JsonClass curjcls;
	private JsonField curjfld;

	private JsonEncoder(JsonEncoder parent, Object object) throws Exception {
		par = parent;
		obj = object;
		encode(obj.getClass());
	}

	private void encode(Class<?> cls) throws Exception {
		if (cls.getSuperclass().getAnnotation(JsonClass.class) != null)
			encode(cls.getSuperclass());
		curjcls = cls.getAnnotation(JsonClass.class);
		for (Field f : cls.getDeclaredFields())
			if (curjcls.noTag() == JsonClass.NoTag.LOAD || f.getAnnotation(JsonField.class) != null) {
				if (Modifier.isStatic(f.getModifiers()))
					continue;
				JsonField jf = f.getAnnotation(JsonField.class);
				if (jf == null)
					jf = JsonField.DEF;
				if (jf.block() || jf.io() == JsonField.IOType.R)
					continue;
				String tag = jf.tag().length() == 0 ? f.getName() : jf.tag();
				f.setAccessible(true);
				ans.add(tag, encode(f.get(obj), getInvoker()));
			}
		for (Method m : cls.getDeclaredMethods())
			if (m.getAnnotation(JsonField.class) != null) {
				JsonField jf = m.getAnnotation(JsonField.class);
				if (jf.io() == JsonField.IOType.R)
					continue;
				if (jf.io() == JsonField.IOType.RW)
					throw new JsonException(Type.FUNC, null, "functional fields should not have RW type");
				String tag = jf.tag();
				if (tag.length() == 0)
					throw new JsonException(Type.TAG, null, "function fields must have tag");
				ans.add(tag, encode(m.invoke(obj), getInvoker()));
			}
	}

	private JsonEncoder getInvoker() {
		return curjcls.bypass() ? par : this;
	}

}

File: io/json/JsonEncoder.kt
package common.io.json

import common.io.json.JsonField.IOType
import common.io.json.JsonField.SerType
import common.pack.Context.SupExc
import common.util.Data
import java.lang.Exceptionimport
import java.lang.reflect.Array
import java.lang.reflect.Modifier

com.google.api.client.json.jackson2.JacksonFactory
import kotlin.Throws
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import common.io.json.JsonClass.NoTag
import common.io.json.JsonClass.WType
import com.google.gson.JsonArray
import common.io.json.JsonClass.JCGeneric
import com.google.gson.JsonPrimitive
import com.google.gson.JsonNull
import common.io.json.JsonClass.JCIdentifier

class JsonEncoder private constructor(private val par: JsonEncoder?, private val obj: Any) {
    private val ans: JsonObject = JsonObject()
    private var curjcls: JsonClass? = null
    private var curjfld: JsonField? = null
    private var index = 0
    @Throws(Exception::class)
    private fun encode(cls: Class<*>) {
        if (cls.superclass.getAnnotation(JsonClass::class.java) != null) encode(cls.superclass)
        curjcls = cls.getAnnotation(JsonClass::class.java)
        for (f in cls.declaredFields) if (curjcls.noTag() == NoTag.LOAD || f.getAnnotation(JsonField::class.java) != null) {
            if (Modifier.isStatic(f.modifiers)) continue
            var jf: JsonField = f.getAnnotation(JsonField::class.java)
            if (jf == null) jf = JsonField.Companion.DEF
            if (jf.block() || jf.io() == IOType.R) continue
            val tag = if (jf.tag().length == 0) f.name else jf.tag()
            f.isAccessible = true
            curjfld = jf
            ans.add(tag, encode(f[obj], getInvoker()))
            curjfld = null
        }
        for (m in cls.declaredMethods) if (m.getAnnotation(JsonField::class.java) != null) {
            val jf: JsonField = m.getAnnotation(JsonField::class.java)
            if (jf.io() == IOType.R) continue
            if (jf.io() == IOType.RW) throw JsonException(JsonException.Type.FUNC, null, "functional fields should not have RW type")
            val tag: String = jf.tag()
            if (tag.length == 0) throw JsonException(JsonException.Type.TAG, null, "function fields must have tag")
            curjfld = jf
            ans.add(tag, encode(m.invoke(obj), getInvoker()))
            curjfld = null
        }
    }

    private fun getInvoker(): JsonEncoder? {
        return if (curjcls!!.bypass()) par else this
    }

    companion object {
        fun encode(obj: Any?): JsonElement {
            return Data.Companion.err<JsonElement>(SupExc<JsonElement> { encode(obj, null) })
        }

        @Throws(Exception::class)
        private fun encode(obj: Any?, par: JsonEncoder?): JsonElement {
            if (obj == null) return JsonNull.INSTANCE
            if (obj is JsonElement) return obj
            if (obj is Number) return JsonPrimitive(obj as Number?)
            if (obj is Boolean) return JsonPrimitive(obj as Boolean?)
            if (obj is String) return JsonPrimitive(obj as String?)
            if (obj is Class<*>) return JsonPrimitive(obj.name)
            val cls: Class<*> = obj.javaClass
            if (cls.isArray) {
                if (par != null && par.curjfld != null && par.curjfld.usePool()) {
                    val handler = JsonField.Handler()
                    val n = Array.getLength(obj)
                    val jarr = JsonArray()
                    for (i in 0 until n) jarr.add(handler.add(Array.get(obj, i)))
                    val jobj = JsonObject()
                    jobj.add("pool", encode(handler.list))
                    jobj.add("data", jarr)
                    return jobj
                }
                val n = Array.getLength(obj)
                val arr = JsonArray(n)
                for (i in 0 until n) arr.add(encode(Array.get(obj, i), par))
                return arr
            }
            if (cls.getAnnotation(JCGeneric::class.java) != null && par != null && par.curjfld!!.alias().size > par.index) {
                val jcg: JCGeneric = cls.getAnnotation(JCGeneric::class.java)
                val alias: Class<*> = par.curjfld!!.alias().get(par.index)
                var found = false
                for (ala in jcg.value()) if (ala == alias) {
                    found = true
                    break
                }
                if (!found) throw JsonException(JsonException.Type.TYPE_MISMATCH, null, "class not present in JCGeneric")
                for (f in cls.declaredFields) {
                    val jcgw: JCIdentifier = f.getAnnotation(JCIdentifier::class.java)
                    if (jcgw != null && f.type == alias) return encode(f[obj], par)
                }
                val con = alias.getConstructor(cls)
                return encode(con.newInstance(obj), par)
            }
            if (par != null && par.curjfld != null) {
                val jfield: JsonField? = par.curjfld
                if (jfield!!.ser() == SerType.FUNC) {
                    if (jfield.serializer().length == 0) throw JsonException(JsonException.Type.FUNC, null, "no serializer function")
                    val m = par.obj.javaClass.getMethod(jfield.serializer(), cls)
                    return encode(m.invoke(par.obj, obj))
                } else if (jfield.ser() == SerType.CLASS) {
                    val cjc: JsonClass = cls.getAnnotation(JsonClass::class.java)
                    if (cjc == null || cjc.serializer().length == 0) throw JsonException(JsonException.Type.FUNC, null, "no serializer function")
                    val func: String = cjc.serializer()
                    val m = cls.getMethod(func)
                    return encode(m.invoke(obj), null)
                }
            }
            val jc: JsonClass = cls.getAnnotation(JsonClass::class.java)
            if (jc != null) if (jc.write() == WType.DEF) return JsonEncoder(par, obj).ans else if (jc.write() == WType.CLASS) {
                if (jc.serializer().length == 0) throw JsonException(JsonException.Type.FUNC, null, "no serializer function")
                val func: String = jc.serializer()
                val m = cls.getMethod(func)
                return encode(m.invoke(obj), null)
            }
            if (obj is List<*>) return encodeList(obj, par)
            if (obj is Set<*>) return encodeSet(obj, par)
            if (obj is Map<*, *>) return encodeMap(obj, par)
            throw JsonException(JsonException.Type.UNDEFINED, null, "object $obj not defined")
        }

        @Throws(Exception::class)
        private fun encodeList(list: List<*>, par: JsonEncoder?): JsonElement {
            if (par != null && par.curjfld != null && par.curjfld.usePool()) {
                val handler = JsonField.Handler()
                val n = list.size
                val jarr = JsonArray()
                for (i in 0 until n) jarr.add(handler.add(list[i]))
                val jobj = JsonObject()
                jobj.add("pool", encode(handler.list))
                jobj.add("data", jarr)
                return jobj
            }
            val ans = JsonArray(list.size)
            for (obj in list) ans.add(encode(obj, par))
            return ans
        }

        @Throws(Exception::class)
        private fun encodeMap(map: Map<*, *>, par: JsonEncoder?): JsonArray {
            val ans = JsonArray(map.size)
            for ((key, value) in map) {
                val ent = JsonObject()
                ent.add("key", encode(key, par))
                par!!.index = 1
                ent.add("val", encode(value, par))
                par.index = 0
                ans.add(ent)
            }
            return ans
        }

        @Throws(Exception::class)
        private fun encodeSet(set: Set<*>, par: JsonEncoder?): JsonArray {
            val ans = JsonArray(set.size)
            for (obj in set) ans.add(encode(obj, par))
            return ans
        }
    }

    init {
        encode(obj.javaClass)
    }
}
